plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelsAction.java

```diff
@@ -116,12 +116,12 @@ private void validateAccess(String modelId, ActionListener<Boolean> listener) {
         User user = RestActionUtils.getUserContext(client);
         String[] excludes = new String[] { MLModel.MODEL_CONTENT_FIELD, MLModel.OLD_MODEL_CONTENT_FIELD };
         try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
-            mlModelManager.getModel(modelId, null, excludes, ActionListener.wrap(mlModel -> {
+            mlModelManager.getModel(modelId, null, excludes, ActionListener.runBefore(ActionListener.wrap(mlModel -> {
                 modelAccessControlHelper.validateModelGroupAccess(user, mlModel.getModelGroupId(), client, listener);
             }, e -> {
                 log.error("Failed to find Model", e);
                 listener.onFailure(e);
-            }));
+            }), () -> context.restore()));
```

Comment:
Isn't the context restored anyway after this statement?

Reply:
yes, but the action listener is running inside it